components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/NamedJdbcTemplate.java

```diff
@@ -0,0 +1,401 @@
+            if (transactionEntry.getTransactionDepth() == 0) {
+                try {
+                    transactionEntry.getConnection().rollback();
+                } catch (SQLException e) {
+                    logDebug("Could not rollback the transaction.", e);
+                    throw new TransactionException(JdbcConstants.ErrorCodes.ERROR_CODE_ROLLBACK_TRANSACTION_FAILURE.
+                            getErrorMessage(), JdbcConstants.ErrorCodes.ERROR_CODE_ROLLBACK_TRANSACTION_FAILURE.
+                            getErrorCode(), e);
+                }
+            }
+            throw new TransactionException(JdbcConstants.ErrorCodes.ERROR_CODE_COMMIT_TRANSACTION_FAILURE.
+                    getErrorMessage(), JdbcConstants.ErrorCodes.ERROR_CODE_COMMIT_TRANSACTION_FAILURE.
+                    getErrorCode(), t);
+        } finally {
+            if (transactionEntry.getTransactionDepth() == 0) {
+                try {
+                    transactionEntry.getConnection().setAutoCommit(transactionEntry.getAutoCommitState());
+                    transactionEntry.getConnection().close();
+                    transactionEntry.setConnection(null);
+                } catch (SQLException e) {
+                    logDebug("Could not close the transaction.", e);
+                    throw new TransactionException(JdbcConstants.ErrorCodes.ERROR_CODE_CLOSE_CONNECTION_FAILURE.
+                            getErrorMessage(), JdbcConstants.ErrorCodes.ERROR_CODE_CLOSE_CONNECTION_FAILURE.
+                            getErrorCode(), e);
+                }
+                TransactionManager.exitTransaction();
+            }
+        }
+    }
+
+    /**
+     * Executes a query on JDBC and return the result as a list of domain objects.
+     *
+     * @param query     the SQL query with the parameter placeholders.
+     * @param rowMapper Row mapper functional interface.
+     * @return List of domain objects of required type.
+     * @see #executeQuery(String, RowMapper, NamedQueryFilter)
+     */
+    public <T> List<T> executeQuery(String query, RowMapper<T> rowMapper) throws
+            DataAccessException {
+
+        return executeQuery(query, rowMapper, null);
+    }
+
+    /**
+     * Executes a query on JDBC and return the result as a list of domain objects.
+     *
+     * @param query            the SQL query with the parameter placeholders.
+     * @param rowMapper        Row mapper functional interface.
+     * @param namedQueryFilter parameters for the SQL query parameter replacement.
+     * @return List of domain objects of required type.
+     */
+    public <T> List<T> executeQuery(String query, RowMapper<T> rowMapper, NamedQueryFilter namedQueryFilter)
+            throws DataAccessException {
+
+        List<T> result = new ArrayList<>();
+        try (Connection connection = dataSource.getConnection()) {
+            NamedPreparedStatement namedPreparedStatement = new NamedPreparedStatement(connection, query);
```

Comment:
This looks a puzzel to me. I would expect NamedPreparedStatement to behave as prepared statements. ie. The user of the NamedPreparedStatment should not see that internally in NamedPreparedStatement we have used a prepared statement. We should avoid exposing and dealing with underlying prepared statement at outer layers. Please clarify if I am overlooking something.

Reply:
Will change the NamedPreparedStatement class (by implementing the PreparedStatement interface) to avoid exposing of the prepared statement.